discovery/sync_manager.go

```diff
@@ -0,0 +1,640 @@
+		case <-m.cfg.HistoricalSyncTicker.Ticks():
+			m.forceHistoricalSync()
+
+		// A new activeSyncer has been received, so we'll either attempt
+		// to start it if the queue is empty or queue it so that we
+		// don't attempt to sync with multiple peers at the same time.
+		case s := <-m.newActiveSyncers:
+			if current == nil {
+				if err := startSyncer(s); err != nil {
+					log.Errorf("Unable to start "+
+						"GossipSyncer(%x): %v", err)
+					continue
+				}
+
+				current = s
+				continue
+			}
+
+			log.Debugf("Queueing next active syncer for peer=%x",
+				s.cfg.peerPub)
+
+			m.pendingActiveSyncers[s.cfg.peerPub] = s
+			m.pendingActiveSyncerQueue.PushBack(s)
+
+		// A stale activeSyncer has been received, so we'll need to
+		// remove them from our queue. If we are currently waiting for
+		// its state machine to complete, we'll move on to the next
+		// activeSyncer in the queue.
+		case peer := <-m.staleActiveSyncers:
+			if current != nil && peer == current.cfg.peerPub {
+				startNextSyncer()
+				continue
+			}
+
+			delete(m.pendingActiveSyncers, peer)
+
+		// Our current active syncer has reached its terminal
+		// chansSynced state, so we'll proceed to starting the next
+		// pending active syncer if there is one.
+		case <-startNext:
+			startNextSyncer()
+
+		// We've timed out waiting for the current active syncer to
+		// reach its terminal chansSynced state.
+		case <-m.cfg.ActiveSyncerTimeoutTicker.Ticks():
+			// If there are no pending active syncers left, then we
+			// can just reset the timeout for the current one.
+			if len(m.pendingActiveSyncers) == 0 {
+				// TODO(wilmer): Restart syncer state machine?
+				log.Debugf("Rescheduling active syncer timeout "+
+					"for peer=%x", current.cfg.peerPub)
+				continue
+			}
+
+			// Otherwise, we'll move on to the next active syncer in
+			// the list and add the current one to the end of our
+			// queue.
+			log.Debugf("Timed out waiting for active syncer with "+
+				"peer=%x to complete", current.cfg.peerPub)
+
+			m.pendingActiveSyncers[current.cfg.peerPub] = current
+			m.pendingActiveSyncerQueue.PushBack(current)
+
+			startNextSyncer()
+
+		case <-m.quit:
+			return
+		}
+	}
+}
+
+// nextPendingActiveSyncer returns the next activeSyncer pending to be started.
+// If there aren't any, then `nil` is returned.
+func (m *SyncManager) nextPendingActiveSyncer() *activeSyncer {
+	next := m.pendingActiveSyncerQueue.Front()
+	for next != nil {
+		s := m.pendingActiveSyncerQueue.Remove(next).(*activeSyncer)
+
+		// If the next pending active syncer is no longer in our lookup
+		// map, then the corresponding peer has disconnected, so we'll
+		// skip them.
+		if _, ok := m.pendingActiveSyncers[s.cfg.peerPub]; !ok {
+			next = m.pendingActiveSyncerQueue.Front()
+			continue
+		}
+
+		return s
+	}
+
+	return nil
+}
+
+// rotateActiveSyncers rotates all currently eligible active syncers.
+func (m *SyncManager) rotateActiveSyncers() {
```

Comment:
Guess this should be named `rotateActiveSyncerCandidates` or similar, current name a bit confusing.

Reply:
Fixed.